cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,234 @@
+    Returns the name of the current WSL instance as seen from outside.
+    """
+    root_net_path = wsl_path_2_win("/")
+    return root_net_path.name
+
+
+def mounted_win_drives() -> List[str]:
+    """
+    Return a list of mount points of the Windows drives inside the current
+    WSL instance, if drives are mounted, or an empty list otherwise
+    """
+    FS_TYPE = "9p"
+    OPTIONS_CONTAIN = "aname=drvfs"
+
+    mounted = []
+    for mnt in util.mounts().values():
+        if mnt["fstype"] == FS_TYPE and OPTIONS_CONTAIN in mnt["opts"]:
+            mounted.append(mnt["mountpoint"])
+
+    return mounted
+
+
+def win_path_2_wsl(path: str) -> PurePath:
+    """
+    Returns a translation of a Windows path to a Linux path that can be
+    accessed inside the current instance filesystem.
+
+    It requires the Windows drive mounting feature to be enabled and the
+    disk drive must be muonted for this to succeed.
+
+    Example:
+    # Assuming Windows drives are mounted under /mnt/ and "S:" doesn't exist:
+    p = winpath2wsl("C:\\ProgramData") # p == "/mnt/c/ProgramData/"
+    n = winpath2wsl("S:\\CoolFolder") # Exception! S: is not mounted.
+
+    :param path: string representing a Windows path. The root drive must exist,
+    although the path is not required to.
+    """
+    out, _ = subp.subp([WSLPATH_CMD, "-au", path])
+    return PurePath(out.rstrip())
+
+
+def cmd_executable() -> PurePath:
+    """
+    Returns the Linux path to the Windows host's cmd.exe.
+    """
+
+    mounts = mounted_win_drives()
+    if not mounts:
+        raise IOError("Windows drives are not mounted.")
+
+    # cmd.exe path is being stable for decades.
+    candidate = "%s/Windows/System32/cmd.exe"
+    for mnt in mounts:
+        cmd = candidate % mnt
+        if not os.access(cmd, os.X_OK):
+            continue
+
+        LOG.debug("Found cmd.exe at <%s>", cmd)
+        return PurePath(cmd)
+
+    raise IOError(
+        "Couldn't find cmd.exe in any mount point: %s" % ", ".join(mounts)
+    )
+
+
+def win_user_profile_dir() -> PurePath:
+    """
+    Returns the Windows user profile directory translated as a Linux path
+    accessible inside the current WSL instance.
+    """
+    cmd = cmd_executable()
+
+    # cloud-init runs too early to rely on binfmt to execute Windows binaries.
+    # But we know that `/init` is the interpreter, so we can run it directly.
+    # See /proc/sys/fs/binfmt_misc/WSLInterop[-late]
+    # inside any WSL instance for more details.
+    home, _ = subp.subp(["/init", cmd.as_posix(), "/C", "echo %USERPROFILE%"])
+    home = home.rstrip()
+    if not home:
+        raise subp.ProcessExecutionError(
+            "No output from cmd.exe to show the user profile dir."
+        )
+
+    return win_path_2_wsl(home.rstrip())
+
+
+def machine_id():
+    """
+    Returns the local machine ID value from /etc/machine-id.
+    """
+    from cloudinit.settings import MACHINE_ID_FILE
+
+    if util.wait_for_files([MACHINE_ID_FILE], 2.0):
```

Comment:
We can't just wait for the existence of this file - it might say "uninitialized" and not yet have a unique id. That means that a subsequent boot would unexpectedly re-run cloud-init.

Reply:
To work reliably, we should probably order cloud-init after [first-boot-complete.target](https://www.freedesktop.org/software/systemd/man/latest/systemd.special.html#first-boot-complete.target).

Reply:
Must that ordering be specific to WSL? I mean, for other use cases under systemd, would it be fine to have that ordering applied?

Reply:
I'd be hesitant adding general systemd ordering changes in order to support this work without us providing a more thorough specification for review to get input on the impacts on other distros/clouds when /etc/machine-id changes  as well as impacts to boot speed and boot ordering cycles on any systemd environment we currently support. Maybe it makes sense to drop this reading of /etc/machine-id as instance-id from WSL, since WSL already has a unique identifier per user which should be the instance name.

The assertions I'd like confirmation on:
1.  A single user cannot have two WSL instances with the same instance name
2. If a WSL instance or snapshot or image gets its name changed, we want cloud-init to re-run
3. The only way to trigger cloud-init to re-run and apply user-data again would be to rename the instance and reboot

If the above are true, and we don't expect a use case where WSL instances to would need trigger cloud-init to re-run on the same instance name, I don't think we need to build in a dependency on /etc/machine-id for this datasource

Reply:
> 1. A single user cannot have two WSL instances with the same instance name

That's true.
![image](https://github.com/canonical/cloud-init/assets/11138291/6fde7735-e487-4c32-a6e0-0945953ca56b)

> 2. If a WSL instance or snapshot or image gets its name changed, we want cloud-init to re-run

There is no convenient interface at the moment to rename an instance (either from inside or outside), but that's possible through the Windows registry:

![image](https://github.com/canonical/cloud-init/assets/11138291/ff9b516f-cabe-4f7c-a738-cf66acc333c2)

Having the datasource recomputing and comparing the id with the cached value can ensure we react to the instance name changing.

> 3. The only way to trigger cloud-init to re-run and apply user-data again would be to rename the instance and reboot

The discussion around `/etc/machine-id` shows that it's quite brittle to rely on it, since it's not universaly available for all Linux distros and the timing around its generation might also be an issue.

So, @blackboxsw and myself brainstormed about this and come to a conclusion that the machine ID must come from the outside, being dependent on:
1. **The Windows host**. An image exported from a computer and imported in another computer would detect it's a new fresh instance and trigger cloud-init to re-run.
2. **The Windows user**. An image exported to a location in disk where other users can access and create new instances imported from that image wouldn't fool cloud-init, i.e. it would detect those new instances as new and re-run.
3. **Some internal knowledge WSL has about the instance**. A user could export a snapshot of a distro instance, unregister it - delete the instance - and reimport the instance under the same name. None of the pieces of information above would be enough to detect that it's a new instance. But WSL internally stores UUID's in the Windows registry. Even in the case mentioned above the new instance receives a new UUID. We cannot ensure how unique is that UUID in a broader scope, but that value combined with the other pieces of information already mentioned allows us to compute a quite unique ID for the instance.

So, besides implementing a method `check_instance_id()` for this datasource, I need to compute the that ID differently, by launching some nasty Windows commands via `subp` to acquire the following information:

1. **The Windows host**: `powershell.exe -Command '$(Get-WmiObject -Class "Win32_ComputerSystemProduct").UUID'` => UUID tied to the Windows hardware
2. **The Windows user**: `powershell.exe -Command 'whoami.exe'` => we're not gonna store it, but hash it soon
3. **Some internal knowledge WSL has about the instance**: `powershell.xe -Command "reg.exe query 'HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Lxss\'" => We can find the UUID for the current instance looking inside this registry location.

Since the user can disable access to those binaries anytime after the first boot, if the datasource fails to compute the current instance ID due lack of access, we'd assume nothing has changed.

I think that summarizes the brainstorming we did.

Reply:
@CarlosNihelton Great investigation - thanks for looking into the options.

After our discussion today about using the instance name and instance-id provided by the metadata file, @blackboxsw and @TheRealFalcon  and I talked a little more. Since it seems rather arbitrary that instance name change would retrigger cloud-init, I think we're leaning towards using a known default instance id ([as in the case of nocloud](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceNoCloud.py#L53)) unless the user provides one in meta-data.

This solution should be very simple, and hopefully more robust than calling into powershell and more obvious to users of cloud-init than using instance name as a trigger for whether cloud-init should run.

Reply:
Thanks @CarlosNihelton for the continuation of this discussion. You referenced in another discussion that if cloud-init relies on powershell functionality such as Get_WmiObject, whoami or reg.exe, a user could disable access to any of those commands on the host, When that config visibility is disabled, it would leave cloud-init with the inability to determine instance-id on those machines. Since that configuration change is security-related and may become policy for an organization, we determined it is best to avoid such commands as they may be ineffective in some configurations leaving cloud-init in a state where cannot determine instance-id across a reboot boundary.

As a result, we decided the WSL datasource can instead rely on setting a default `instance-id: datasource-wsl` in metadata, but allow for reading an optional `<instance-name>`.meta-data file on the Windows host can be provided alongside the applicable `.user-data`. If `<instance-name>.meta-data` exists, it could contain a JSON dicttionary optionally providing a some value for instance id such as:  `'{"instance-id": "x-y-z"}'`. As with other cloud platforms, that optional meta-data file can change the value of instance-id to trigger cloud-init to re-run across reboot. If users are looking to share snapshots or relaunch a snapshot where cloud-init is re-triggered, they would either run `sudo cloud-init clean --logs`  on the instance before snapshot/export, or create the appropriate <new-instance-name>.meta-data file containing  `'{"instance-id": "some-new-instance-id"}'`.